codegen/src/main/java/software/amazon/awssdk/codegen/poet/rules2/CodeGeneratorVisitor.java

```diff
@@ -274,7 +278,11 @@ private void codegenTreeBody(RuleSetExpression expr) {
     @Override
     public Void visitEndpointExpression(EndpointExpression e) {
         builder.add("return $T.endpoint(", typeMirror.rulesResult().type());
-        builder.add("$T.builder().url($T.create(", Endpoint.class, URI.class);
+        if (endpointCaching) {
+            builder.add("$T.builder().url($T.getInstance().create(", Endpoint.class, SdkUri.class);
```

Comment:
Can we add a codegen tests? 

Actually, should we just do it for all service clients at this point? I don't really see us removing the logic ever...

Reply:
I would avoid doing it for all service clients to minimize blast radius. But good point on codegen test, I'll ad them.